igvm/src/c_api.rs

```diff
@@ -0,0 +1,763 @@
+    HEADER_SECTION_INITIALIZATION,
+    HEADER_SECTION_DIRECTIVE,
+}
+
+fn translate_error(error: Error) -> IgvmApiError {
+    match error {
+        Error::NoPlatformHeaders => IgvmApiError::IGVMAPI_NO_PLATFORM_HEADERS,
+        Error::FileDataSectionTooLarge => IgvmApiError::IGVMAPI_FILE_DATA_SECTION_TOO_LARGE,
+        Error::VariableHeaderSectionTooLarge => {
+            IgvmApiError::IGVMAPI_VARIABLE_HEADER_SECTION_TOO_LARGE
+        }
+        Error::TotalFileSizeTooLarge => IgvmApiError::IGVMAPI_TOTAL_FILE_SIZE_TOO_LARGE,
+        Error::InvalidBinaryPlatformHeader(_) => {
+            IgvmApiError::IGVMAPI_INVALID_BINARY_PLATFORM_HEADER
+        }
+        Error::InvalidBinaryInitializationHeader(_) => {
+            IgvmApiError::IGVMAPI_INVALID_BINARY_INITIALIZATION_HEADER
+        }
+        Error::InvalidBinaryDirectiveHeader(_) => {
+            IgvmApiError::IGVMAPI_INVALID_BINARY_DIRECTIVE_HEADER
+        }
+        Error::MultiplePlatformHeadersWithSameIsolation => {
+            IgvmApiError::IGVMAPI_MULTIPLE_PLATFORM_HEADERS_WITH_SAME_ISOLATION
+        }
+        Error::InvalidParameterAreaIndex => IgvmApiError::IGVMAPI_INVALID_PARAMETER_AREA_INDEX,
+        Error::InvalidPlatformType => IgvmApiError::IGVMAPI_INVALID_PLATFORM_TYPE,
+        Error::NoFreeCompatibilityMasks => IgvmApiError::IGVMAPI_NO_FREE_COMPATIBILITY_MASKS,
+        Error::InvalidFixedHeader => IgvmApiError::IGVMAPI_INVALID_FIXED_HEADER,
+        Error::InvalidBinaryVariableHeaderSection => {
+            IgvmApiError::IGVMAPI_INVALID_BINARY_VARIABLE_HEADER_SECTION
+        }
+        Error::InvalidChecksum {
+            expected: _,
+            header_value: _,
+        } => IgvmApiError::IGVMAPI_INVALID_CHECKSUM,
+        Error::MultiplePageTableRelocationHeaders => {
+            IgvmApiError::IGVMAPI_MULTIPLE_PAGE_TABLE_RELOCATION_HEADERS
+        }
+        Error::RelocationRegionsOverlap => IgvmApiError::IGVMAPI_RELOCATION_REGIONS_OVERLAP,
+        Error::ParameterInsertInsidePageTableRegion => {
+            IgvmApiError::IGVMAPI_PARAMETER_INSERT_INSIDE_PAGE_TABLE_REGION
+        }
+        Error::NoMatchingVpContext => IgvmApiError::IGVMAPI_NO_MATCHING_VP_CONTEXT,
+        Error::PlatformArchUnsupported {
+            arch: _,
+            platform: _,
+        } => IgvmApiError::IGVMAPI_PLATFORM_ARCH_UNSUPPORTED,
+        Error::InvalidHeaderArch {
+            arch: _,
+            header_type: _,
+        } => IgvmApiError::IGVMAPI_INVALID_HEADER_ARCH,
+        Error::UnsupportedPageSize(_) => IgvmApiError::IGVMAPI_UNSUPPORTED_PAGE_SIZE,
+        Error::InvalidFixedHeaderArch(_) => IgvmApiError::IGVMAPI_INVALID_FIXED_HEADER_ARCH,
+        Error::MergeRevision => IgvmApiError::IGVMAPI_MERGE_REVISION,
+    }
+}
+
+type IgvmResult = i64;
+type IgvmHandle = i64;
+
+struct IgvmFileInstance {
+    file: IgvmFile,
+    buffers: BTreeMap<IgvmHandle, Vec<u8>>,
+}
+
+static mut IGVM_HANDLES: Mutex<BTreeMap<IgvmHandle, IgvmFileInstance>> =
```

Comment:
I think you can remove the need to make this mutable by using std::sync::OnceLock and initializing in the igvm_create call. I would consider it an error to call the other API methods anyways since you don't have a handle. 
See: https://doc.rust-lang.org/std/sync/struct.OnceLock.html 

For example, replace it with:
```rust
static IGVM_HANDLES: OnceLock<Mutex<BTreeMap<...>
``` 

Then the corresponding usage in `igvm_create` can be modified to remove the unsafe around mutable usage to:
```rust
let handle = IGVM_HANDLE_FACTORY.get_or_init(|| AtomicI64::new(1)).fetch_add(1, Ordering::Relaxed);
let mut m = IGVM_HANDLES.get_or_init(|| Mutex::new(BTreeMap::new());
...
```

And other usages of the globals can be done with `IGVM_HANDLES.get().expect("caller called API before igvm_create")`.

Which should help us get rid of the unsafe blocks and the need for SAFETY comments.

Reply:
Or you could return some kind of Invalid handle error also if the `.get()` fails. Maybe that's better?

Reply:
Maybe its worth abstracting additionally the IGVM_HANDLES handout function to a rust function as well, so you can do the `get_or_init` call and also save having to copy-paste the `add(1, Ordering::Relaxed)` everywhere, too.

Reply:
I've implemented your suggestions and return an error code if `.get()` fails. This is much cleaner. Thanks.